### PR TITLE
4.x internal filesystem

### PR DESCRIPTION
**Pull request by ADmad**
Refs #12569.

My previous PR #12650 combined with this takes care of removing all usage of `File` and `Folder` from core. I'll remove it from tests too once this approach is approved.

I have added a new internal class `Cake\Filesystem\Filesystem` with utility methods for file system operations to avoid code repetition.

### CODE REVIEW

File: src/Console/CommandScanner.php [hunk 1]
```diff
@@ -109,41 +109,40 @@ public function scanPlugin(string $plugin): array
...
-            $shells[] = [
+            $shells[$path . $file] = [
                 'file' => $path . $file,
                 'fullName' => $prefix . $name,
                 'name' => $name,
                 'class' => $class,
             ];
         }
 
+        ksort($shells);
```
Comment (markstory) [inline src/Console/CommandScanner.php]: Should you `array_values()` after this?

File: src/Filesystem/Filesystem.php [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+        $iterator = new RecursiveIteratorIterator(
+            new RecursiveDirectoryIterator($path, FilesystemIterator::SKIP_DOTS),
+            RecursiveIteratorIterator::CHILD_FIRST
+        );
+
+        $result = true;
+        foreach ($iterator as $fileInfo) {
+            switch ($fileInfo->getType()) {
+                case 'dir':
```
Comment (markstory) [inline src/Filesystem/Filesystem.php]: Does PHP give us constants for these strings?

Comment (ADmad) [reply src/Filesystem/Filesystem.php]: Not that I know of, will check.

File: src/Console/CommandScanner.php (hunk 1 above)
Comment (ADmad) [reply src/Console/CommandScanner.php]: We could but not really necessary, the keys are never used.

File: src/Filesystem/Filesystem.php (hunk 2 above)
Comment (garas) [reply src/Filesystem/Filesystem.php]: Add `RecursiveDirectoryIterator::CURRENT_AS_SELF` flag to `RecursiveDirectoryIterator` constructor. 

`$fileInfo` is becomes instance of `RecursiveDirectoryIterator`, and has 

-    `DirectoryIterator::isDir()` - Determine if current DirectoryIterator item is a directory
-    `DirectoryIterator::isDot()` - Determine if current DirectoryIterator item is '.' or '..'
-    `DirectoryIterator::isFile()` - Determine if current DirectoryIterator item is a regular file
-    `DirectoryIterator::isLink()` - Determine if current DirectoryIterator item is a symbolic link

File: src/Filesystem/Filesystem.php [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+        return new CallbackFilterIterator($flatten, $filter);
+    }
+
+    /**
+     * Dump contents to file.
+     *
+     * @param string $filename File path.
+     * @param string $content Content to dump.
+     * @return void
+     * @throws \Cake\Core\Exception When dumping fails.
```
Comment (garas) [inline src/Filesystem/Filesystem.php]: this and 2 more in this file: `\Cake\Core\Exception\Exception`

File: src/Filesystem/Filesystem.php [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+    /**
+     * Find files (non-recursively) in given directory path.
+     *
+     * @param string $path Directory path.
+     * @param mixed $filter If string will be used as regex for filtering using
+     *   `RegexIterator`, if callable will be as callback for `CallbackFilterIterator`.
+     * @param int|null $flags Flags for FilesystemIterator::__construct();
+     * @return \Traversable
+     */
+    public function find(string $path, $filter = null, ?int $flags = null): Traversable
```
Comment (inoas) [inline src/Filesystem/Filesystem.php]: `findFile()`? If ti finds directories and symbolic links as well... the docblock is wrong?

File: src/Filesystem/Filesystem.php [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+     * Find files recursively in given directory path.
+     *
+     * @param string $path Directory path.
+     * @param mixed $filter If string will be used as regex for filtering using
+     *   `RegexIterator`, if callable will be as callback for `CallbackFilterIterator`.
+     *   Hidden directories (starting with dot e.g. .git) are always skipped.
+     * @param int|null $flags Flags for FilesystemIterator::__construct();
+     * @return \Traversable
+     */
+    public function findRecursive(string $path, $filter = null, ?int $flags = null): Traversable
```
Comment (inoas) [inline src/Filesystem/Filesystem.php]: findFilesRecursive()?

Comment (inoas) [reply src/Filesystem/Filesystem.php]: As the signature is the same as findFIles could both be called by one proxy method having recursive = true/false as the last arg?

File: src/Filesystem/Filesystem.php [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+    /**
+     * Create directory.
+     *
+     * @param string $dir Directory path.
+     * @param int $mode Octal mode passed to mkdir(). Defaults to 0755.
+     * @return void
+     * @throws \Cake\Core\Exception\Exception When directory creation fails.
+     */
+    public function mkdir(string $dir, int $mode = 0755): void
```
Comment (inoas) [inline src/Filesystem/Filesystem.php]: Why use mkdir but then deleteDir copyDir?
what about either of these for consistency:

mkdir, rmdir, cpdir
mkDir, rmDir, cpDir
makeDir, removeDir, copyDir

**Comment by inoas** [comment]
> I am still in favor of using `symfony/filesystem` instead of the poor man's implementation I have done.

Thanks for doing this, but why was that rejected?

**Comment by markstory** [comment]
I like this approach. It solves the problems we have without adding a non-trivial dependency and the new code looks good.

File: src/Filesystem/Filesystem.php (hunk 6 above)
Comment (ADmad) [reply src/Filesystem/Filesystem.php]: This class is only meant for internal use and won't be documented in the API hence there's no need to be too pedantic about the naming.

File: src/Filesystem/Filesystem.php (hunk 5 above)
Comment (ADmad) [reply src/Filesystem/Filesystem.php]: Boolean args are frowned upon. A more explicit API is better.

**Comment by ADmad** [comment]
phpstan failure fixed on 4.x.

**Comment by ADmad** [comment]
Anything more needs to be done here?

**Comment by markstory** [comment]
I'll take one more look :)

**Pull request merged**